readthedocs/invitations/models.py

```diff
@@ -0,0 +1,209 @@
+        queryset = self.filter(created__gt=self._expiration_days)
+        if obj:
+            queryset = self._for_object(obj=obj, queryset=queryset)
+        return queryset
+
+    def invite(self, from_user, obj, to_user=None, to_email=None):
+        """
+        Create and send an invitation for `to_user` or `to_email` to join `object`.
+
+        If the invitation already exists, we don't send the invitation again.
+        """
+        if not to_user and not to_email:
+            raise ValueError("A user or email must be provided")
+        fields = dict(
+            content_type=ContentType.objects.get_for_model(obj),
+            object_id=obj.pk,
+        )
+        if to_user:
+            fields["to_user"] = to_user
+        else:
+            fields["to_email"] = to_email
+
+        invitation, created = self.get_or_create(
+            **fields,
+            defaults={
+                "from_user": from_user,
+            },
+        )
+        if created:
+            invitation.send()
+        return invitation, created
+
+    def for_object(self, obj):
+        return self._for_object(obj=obj, queryset=self.all())
+
+    @staticmethod
+    def _for_object(obj, queryset):
+        return queryset.filter(
+            object_id=obj.pk,
+            content_type=ContentType.objects.get_for_model(obj),
+        )
+
+
+class Invitation(TimeStampedModel):
+
+    """
+    Invitation model.
+
+    An invitation can be attached to an existing user or to an email.
+    """
+
+    # Generic foreign key.
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    object = GenericForeignKey("content_type", "object_id")
+
+    # Normal fields.
+    from_user = models.ForeignKey(
+        User,
+        verbose_name=_("From user"),
+        on_delete=models.CASCADE,
+        related_name="invitations_sent",
+    )
+    to_user = models.ForeignKey(
+        User,
+        verbose_name=_("To user"),
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+        related_name="invitations_received",
+    )
+    to_email = models.EmailField(_("E-mail"), null=True, blank=True)
+    token = models.CharField(unique=True, max_length=32)
+
+    objects = InvitationQueryset.as_manager()
+
+    class Meta:
+        indexes = [
+            models.Index(fields=["content_type", "object_id"]),
+        ]
+        unique_together = [
+            ("to_user", "content_type", "object_id"),
+            ("to_email", "content_type", "object_id"),
+        ]
+
+    @property
+    def username(self):
+        if self.to_user:
+            return self.to_user.username
+        return self.to_email
+
+    @property
+    def expired(self):
+        return timezone.now() > self.expiration_date
+
+    @property
+    def expiration_date(self):
+        return self.created + timezone.timedelta(
+            days=settings.RTD_INVITATIONS_EXPIRATION_DAYS
+        )
```

Comment:
I like the idea of saving the expiration date in case we change it.

> It also solves a bit of a weird situation where you can't call the property if the database hasn't defined created.

Created is a non-null field, so it's always defined.

Reply:
This is great :+1: I'm sure we can use this field in the future, as we start to expand features in the invitation system.